README.md
# euc

EUC-JP and Shift_JIS Encoding and Decoding Library for Dart Language

[![Build Status](https://travis-ci.org/dsh0416/euc-jp.svg?branch=master)](https://travis-ci.org/dsh0416/euc-jp)

## Examples

```dart
import 'package:euc/euc.dart';
import 'package:euc/jis.dart';

main() {
  // EUC-JP Encoding and Decoding
  print(EucJP().decode([
    164, 170, 164, 207, 164, 232, 164, 166, 192, 164, 179, 166
  ]));

  print(EucJP().encode("おはよう世界"));

  // Shift_JIS Encoding and Decoding
  print(ShiftJIS().decode([
    130, 168, 130, 205, 130, 230, 130, 164, 144, 162, 138, 69
  ]));

  print(ShiftJIS().encode("おはよう世界"));
}
```


CHANGELOG.md
# Releases

- v1.0.0+1 First Implementation
- v1.0.0+2 Add Example and Changelog
- v1.0.1+3 Support Encoder
- v1.0.2+4 Add Shift_JIS Support
- v1.0.3+5 Fix bug in ASCII table
- v1.0.4+6 Remove warning in spec
- v1.0.5+7 Add unit tests
- v1.0.6+8 Fix some missing characters


test/euc.dart
import 'package:test/test.dart';
import 'package:euc-jp/euc.dart';

void main() {
  test('String.split() splits the string on the delimiter', () {
    var string = 'foo,bar,baz';
    expect(string.split(','), equals(['foo', 'bar', 'baz']));
  });

  test('String.trim() removes surrounding whitespace', () {
    var string = '  foo ';
    expect(string.trim(), equals('foo'));
  });
}


test/basic_test.dart
import 'package:test/test.dart';

import 'package:euc/euc.dart';
import 'package:euc/jis.dart';

main() {
  group('EucJP()', () {
    test('.encode()', () {
      expect(EucJP().encode("おはよう世界"),
          equals([164, 170, 164, 207, 164, 232, 164, 166, 192, 164, 179, 166]));
    });

    test('.decode()', () {
      expect(
          EucJP().decode(
              [164, 170, 164, 207, 164, 232, 164, 166, 192, 164, 179, 166]),
          equals("おはよう世界"));
    });
  });

  group('ShiftJIS()', () {
    test('.encode()', () {
      expect(ShiftJIS().encode("おはよう世界"),
          equals([130, 168, 130, 205, 130, 230, 130, 164, 144, 162, 138, 69]));
    });

    test('.decode()', () {
      expect(
          EucJP().decode(
              [164, 170, 164, 207, 164, 232, 164, 166, 192, 164, 179, 166]),
          equals("おはよう世界"));
    });
  });
}


euc_table.rb
euc = {}
utf = {}

# ASCII Compatible
(0x00..0x7E).each do |i|
  euc["0x#{i.to_s(16).rjust(2, '0')}"] = [i].pack('c*').encode("UTF-8", "EUC-JP")
  utf[[i].pack('c*').encode("UTF-8", "EUC-JP").bytes] = [i]
end

# Hiragana
(0xA1..0xDF).each do |i|
  begin
    euc["0x8e#{i.to_s(16).rjust(2, '0')}"] = [0x8e, i].pack('c*').encode("UTF-8", "EUC-JP")
    utf[[0x8e, i].pack('c*').encode("UTF-8", "EUC-JP").bytes] = [0x8e, i]
  rescue Encoding::UndefinedConversionError
    # Ignore
  end
end

# JIS X 0208
(0xA1..0xFE).each do |i|
  (0xA1..0xFE).each do |j|
    begin
      euc["0x#{i.to_s(16).rjust(2, '0')}#{j.to_s(16).ljust(2, '0')}"] = [i, j].pack('c*').encode("UTF-8", "EUC-JP")
      utf[[i, j].pack('c*').encode("UTF-8", "EUC-JP").bytes] = [i, j]
    rescue Encoding::UndefinedConversionError
      # Ignore
    end
  end
end

# JIS X 0212
(0xA1..0xFE).each do |i|
  (0xA1..0xFE).each do |j|
    begin
      euc["0x8f#{i.to_s(16).rjust(2, '0')}#{j.to_s(16).ljust(2, '0')}"] = [0x8F, i, j].pack('c*').encode("UTF-8", "EUC-JP")
      utf[[0x8F, i, j].pack('c*').encode("UTF-8", "EUC-JP").bytes] = [0x8F, i, j]
    rescue Encoding::UndefinedConversionError
      # Ignore
    end
  end
end

puts "const EUC_TABLE = {"
euc.keys.each do |key|
  puts "  #{key}: #{euc[key].unpack('C*')},"
end
puts "};"

puts

puts "const UTF_TABLE = {"
utf.keys.each do |key|
  hex = key.map { |n| '%02x' % (n & 0xFF) }.join
  puts "  0x#{hex}: [#{utf[key].join(', ')}],"
end
puts "};"


lib/euc.dart
import 'dart:convert';
import 'dart:math';
import './euc-table.dart';

class EucJPDecoder extends Converter<List<int>, String> {
  @override
  convert(input) {
    List<int> result = [];
    for (int i = 0; i < input.length; i++) {
      var c1 = input[i];
      if (c1 <= 0x7E) {
        // ASCII Compatible
        result.addAll(EUC_TABLE[c1] ?? []);
      } else if (c1 == 0x8e) {
        // Hiragana
        var c2 = input[++i];
        result.addAll(EUC_TABLE[(c1 << 8) + c2] ?? []);
      } else if (c1 == 0x8f) {
        // JIS X 0212
        var c2 = input[++i];
        var c3 = input[++i];
        result.addAll(EUC_TABLE[(c1 << 16) + (c2 << 8) + c3] ?? []);
      } else {
        // JIS X 0208
        var c2 = input[++i];
        result.addAll(EUC_TABLE[(c1 << 8) + c2] ?? []);
      }
    }
    return utf8.decode(result);
  }
}

class EucJPEncoder extends Converter<String, List<int>> {
  @override
  List<int> convert(String s) {
    List<int> result = [];
    for (int i = 0; i < s.length; i++) {
      var bytes = utf8.encode(s[i]);
      var value = 0;

      for (var i = 0, length = bytes.length; i < length; i++) {
        value += bytes[i] * (pow(256, (bytes.length - i - 1)) as int);
      }

      result.addAll(UTF_TABLE[value] ?? []);
    }
    return result;
  }
}

class EucJP extends Encoding {
  @override
  Converter<List<int>, String> get decoder => EucJPDecoder();

  @override
  Converter<String, List<int>> get encoder => EucJPEncoder();

  @override
  String get name => 'EUC-JP';
}


lib/jis.dart
import 'dart:convert';
import 'dart:math';
import './jis-table.dart';

class JISDecoder extends Converter<List<int>, String> {
  @override
  convert(input) {
    List<int> result = [];
    for (int i = 0; i < input.length; i++) {
      var c1 = input[i];
      if (c1 <= 0x7F) {
        // ASCII Compatible (partially)
        result.addAll(JIS_TABLE[c1] ?? []);
      } else if (c1 >= 0xa1 && c1 <= 0xdf) {
        // Half-width Hiragana
        result.addAll(JIS_TABLE[c1] ?? []);
      } else if (c1 >= 0x81 && c1 <= 0x9f) {
        // JIS X 0208
        var c2 = input[++i];
        result.addAll(JIS_TABLE[(c1 << 8) + c2] ?? []);
      } else if (c1 >= 0xe0 && c1 <= 0xef) {
        // JIS X 0208
        var c2 = input[++i];
        result.addAll(JIS_TABLE[(c1 << 8) + c2] ?? []);
      } else {
        // Unknown
        result.addAll([]);
      }
    }
    return utf8.decode(result);
  }
}

class JISEncoder extends Converter<String, List<int>> {
  @override
  List<int> convert(String s) {
    List<int> result = [];
    for (int i = 0; i < s.length; i++) {
      var bytes = utf8.encode(s[i]);
      var value = 0;

      for (var i = 0, length = bytes.length; i < length; i++) {
        value += bytes[i] * (pow(256, (bytes.length - i - 1)) as int);
      }

      result.addAll(UTF_TABLE[value] ?? []);
    }
    return result;
  }
}

class ShiftJIS extends Encoding {
  @override
  Converter<List<int>, String> get decoder => JISDecoder();

  @override
  Converter<String, List<int>> get encoder => JISEncoder();

  @override
  String get name => 'Shift_JIS';
}


jis_table.rb
jis = {}
utf = {}

# ASCII Compatible
(0x00..0x7F).each do |i|
  jis["0x#{i.to_s(16).rjust(2, '0')}"] = [i].pack('c*').encode("UTF-8", "Shift_JIS")
  utf[[i].pack('c*').encode("UTF-8", "Shift_JIS").bytes] = [i]
end

# Half-width Hiragana
(0xA1..0xDF).each do |i|
  jis["0x#{i.to_s(16).rjust(2, '0')}"] = [i].pack('c*').encode("UTF-8", "Shift_JIS")
  utf[[i].pack('c*').encode("UTF-8", "Shift_JIS").bytes] = [i]
end

# JIS X 0208
(0x81..0x9F).each do |i|
  (0x40..0x7E).each do |j|
    begin
      jis["0x#{i.to_s(16).rjust(2, '0')}#{j.to_s(16).rjust(2, '0')}"] = [i, j].pack('c*').encode("UTF-8", "SJIS")
      utf[[i, j].pack('c*').encode("UTF-8", "Shift_JIS").bytes] = [i, j]
    rescue Encoding::UndefinedConversionError
      # Ignore
    end
  end

  (0x80..0xFC).each do |j|
    begin
      jis["0x#{i.to_s(16).rjust(2, '0')}#{j.to_s(16).rjust(2, '0')}"] = [i, j].pack('c*').encode("UTF-8", "SJIS")
      utf[[i, j].pack('c*').encode("UTF-8", "Shift_JIS").bytes] = [i, j]
    rescue Encoding::UndefinedConversionError
      # Ignore
    end
  end
end


(0xE0..0xEF).each do |i|
  (0x40..0x7E).each do |j|
    begin
      jis["0x#{i.to_s(16).rjust(2, '0')}#{j.to_s(16).rjust(2, '0')}"] = [i, j].pack('c*').encode("UTF-8", "Shift_JIS")
      utf[[i, j].pack('c*').encode("UTF-8", "Shift_JIS").bytes] = [i, j]
    rescue Encoding::UndefinedConversionError
      # Ignore
    end
  end

  (0x80..0xFC).each do |j|
    begin
      jis["0x#{i.to_s(16).rjust(2, '0')}#{j.to_s(16).rjust(2, '0')}"] = [i, j].pack('c*').encode("UTF-8", "Shift_JIS")
      utf[[i, j].pack('c*').encode("UTF-8", "Shift_JIS").bytes] = [i, j]
    rescue Encoding::UndefinedConversionError
      # Ignore
    end
  end
end

puts "const JIS_TABLE = {"
jis.keys.each do |key|
  puts "  #{key}: #{jis[key].unpack('C*')},"
end
puts "};"

puts

puts "const UTF_TABLE = {"
utf.keys.each do |key|
  hex = key.map { |n| '%02x' % (n & 0xFF) }.join
  puts "  0x#{hex}: [#{utf[key].join(', ')}],"
end
puts "};"
